File: demo-service-impl/src/main/java/ru/zubmike/service/demo/dao/db/StarshipDaoImpl.java
package ru.zubmike.service.demo.dao.db;

import org.hibernate.SessionFactory;
import ru.zubmike.service.dao.db.BasicEntityItemDao;
import ru.zubmike.service.demo.dao.StarshipDao;
import ru.zubmike.service.demo.types.Starship;

import javax.inject.Inject;
import java.util.Optional;

public class StarshipDaoImpl extends BasicEntityItemDao<Long, Starship> implements StarshipDao {

	@Inject
	public StarshipDaoImpl(SessionFactory sessionFactory) {
		super(sessionFactory, Starship.class);
	}

	@Override
	public Optional<Starship> getByNumber(String number) {
		return runAndReturn(session -> session.createQuery("from " + clazz.getSimpleName() + " where number = :number",
				Starship.class)
				.setParameter("number", number)
				.uniqueResultOptional());
	}
}

File: demo-service-kotlin-impl/src/main/kotlin/ru/zubmike/service/demo/dao/db/StarshipDaoImpl.kt
package ru.zubmike.service.demo.dao.db

import org.hibernate.SessionFactory
import ru.zubmike.service.dao.db.BasicEntityItemDao
import ru.zubmike.service.demo.dao.StarshipDao
import ru.zubmike.service.demo.types.Starship
import ru.zubmike.service.demo.utils.uniqueResultNullable

class StarshipDaoImpl (sessionFactory: SessionFactory) : BasicEntityItemDao<Long, Starship>(sessionFactory, Starship::class.java), StarshipDao {

    override fun getByNumber(number: String): Starship? {
        return runAndReturn { session ->
            session.createQuery("from " + clazz.simpleName + " where number = :number",
                    Starship::class.java)
                    .setParameter("number", number)
                    .uniqueResultNullable()
        }
    }
}
